src/pdf.py

```diff
@@ -39,86 +45,87 @@ def draw(self, pmlOp, pageNr, output, pe):
 
-            output += "%f w\n"\
-                      "%f %f m\n"\
-                      "%f %f l\n"\
-                      "S\n" % (0.05 * pmlOp.size, x, undY, x + undLen, undY)
+        # create bookmark for table of contents if applicable
+        if pmlOp.toc:
+            bookmarkKey = uuid.uuid4().hex  # we need a unique key to link the bookmark in toc – TODO: generate a more speaking one
+            canvas.bookmarkHorizontal(bookmarkKey, pe.x(pmlOp.x), pe.y(pmlOp.y))
+            canvas.addOutlineEntry(pmlOp.toc.text, bookmarkKey)
 
 class PDFLineOp(PDFDrawOp):
-    def draw(self, pmlOp, pageNr, output, pe):
-        p = pmlOp.points
+    def draw(self, pmlOp: 'pml.DrawOp', pageNr: int, pe: 'PDFExporter', canvas):
+        if not isinstance(pmlOp, pml.LineOp):
+            raise Exception("PDFLineOp is only compatible with pml.LineOp, got "+type(pmlOp).__name__)
 
-        pc = len(p)
+        points = pmlOp.points
+        numberOfPoints = len(points)
 
-        if pc < 2:
-            print("LineOp contains only %d points" % pc)
+        if numberOfPoints < 2:
+            print("LineOp contains only %d points" % numberOfPoints)
 
             return
 
-        output += "%f w\n"\
-                  "%s m\n" % (pe.mm2points(pmlOp.width), pe.xy(p[0]))
+        canvas.setLineWidth(pe.mm2points(pmlOp.width))
 
-        for i in range(1, pc):
-            output += "%s l\n" % (pe.xy(p[i]))
+        lines = []
+        for i in range(0, numberOfPoints - 1):
+            lines.append(pe.xy(points[i]) + pe.xy(points[i+1]))
 
         if pmlOp.isClosed:
-            output += "s\n"
-        else:
-            output += "S\n"
+            lines.append(pe.xy(points[i+1]) + pe.xy(points[0]))
+
+        canvas.lines(lines)
 
 class PDFRectOp(PDFDrawOp):
-    def draw(self, pmlOp, pageNr, output, pe):
+    def draw(self, pmlOp: 'pml.DrawOp', pageNr: int, pe: 'PDFExporter', canvas) -> None:
+        if not isinstance(pmlOp, pml.RectOp):
+            raise Exception("PDFRectOp is only compatible with pml.RectOp, got "+type(pmlOp).__name__)
+
         if pmlOp.lw != -1:
-            output += "%f w\n" % pe.mm2points(pmlOp.lw)
+            canvas.setLineWidth(pe.mm2points(pmlOp.lw))
 
-        output += "%f %f %f %f re\n" % (
+        canvas.rect(
             pe.x(pmlOp.x),
             pe.y(pmlOp.y) - pe.mm2points(pmlOp.height),
-            pe.mm2points(pmlOp.width), pe.mm2points(pmlOp.height))
-
-        if pmlOp.fillType == pml.NO_FILL:
-            output += "S\n"
-        elif pmlOp.fillType == pml.FILL:
-            output += "f\n"
-        elif pmlOp.fillType == pml.STROKE_FILL:
-            output += "B\n"
-        else:
-            print("Invalid fill type for RectOp")
+            pe.mm2points(pmlOp.width),
+            pe.mm2points(pmlOp.height),
+            pmlOp.fillType == pml.NO_FILL or pmlOp.fillType == pml.STROKE_FILL,
+            pmlOp.fillType == pml.FILL or pmlOp.fillType == pml.STROKE_FILL
```

Comment:
The flag `fillType` with the options `NO_FILL`,  `STROKE_FILL` and `FILL` on PML Operation doesn't make so much sense with the reportlab API – 2 seperate boolans for showing the stroke and showing a fill would not only make PML more flexible (and probably easier to understand), it would also save us from this translation from a flag to 2 bools. Maybe this could also be done in this PR.